compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -906,31 +901,24 @@ impl<'tcx, 'a, Tag: Provenance, Extra> AllocRef<'a, 'tcx, Tag, Extra> {
             .alloc
             .read_scalar(&self.tcx, range, read_provenance)
             .map_err(|e| e.to_interp_error(self.alloc_id))?;
-        debug!(
-            "read_scalar in {} at {:#x}, size {}: {:?}",
-            self.alloc_id,
-            range.start.bytes(),
-            range.size.bytes(),
-            res
-        );
+        debug!("read_scalar at {:?}{range:?}: {res:?}", self.alloc_id);
         Ok(res)
     }
 
-    pub fn read_integer(
-        &self,
-        offset: Size,
-        size: Size,
-    ) -> InterpResult<'tcx, ScalarMaybeUninit<Tag>> {
-        self.read_scalar(alloc_range(offset, size), /*read_provenance*/ false)
+    /// `range` is relative to this allocation reference, not the base of the allocation.
+    pub fn read_integer(&self, range: AllocRange) -> InterpResult<'tcx, ScalarMaybeUninit<Tag>> {
+        self.read_scalar(range, /*read_provenance*/ false)
```

Comment:
@oli-obk I know you said you didn't like these `alloc_range(...)` calls. OTOH, I tried changing all these functions to offset-size pairs, and it has the usual problems of a function where two arguments have the same type -- call sites become rather ambiguous-looking:
```rust
                alloc.write_scalar(Size::ZERO, a_size, a_val)?;
```
This is *not* an access of size zero.

By using `alloc_range`, at least one has to only learn once what that function does. I originally intended to use this as `AllocRange { start: Size::ZERO, size: a_size }`, which is very clear but also very verbose.

I am open to other ideas.

Reply:
I recently saw a suggestion for `start..+len`, while neat, I find it an extension to the language that definitely does not carry its weight. The additional sigils aren't very obvious or googleable.

for slices we at least have `x[start..][..len]`. If indexing could return non-references we could go this route. We could use a method, which is likely not worse than what we have:

```rust
        let align = vtable
            .read_integer(alloc_range(
                pointer_size * u64::try_from(COMMON_VTABLE_ENTRIES_ALIGN).unwrap(),
                pointer_size,
            ))?
            .check_init()?;
```

could become


```rust
        let align = vtable
            .offset(pointer_size * u64::try_from(COMMON_VTABLE_ENTRIES_ALIGN).unwrap())
            .with_len(pointer_size)
            .read_integer()?
            .check_init()?;
```

Reply:
fwiw, I think my main complaint is that `read_scalar` and friends take an `AllocRange` argument instead of first calling a method to offset/shrink the alloc and then just having no positional arguments to `read_scalar`

Reply:
> I recently saw a suggestion for start..+len, while neat, I find it an extension to the language that definitely does not carry its weight. The additional sigils aren't very obvious or googleable.

We could have a macro, `alloc_range!(start..+len)`? Though not sure if the parser will allow tat.

> fwiw, I think my main complaint is that read_scalar and friends take an AllocRange argument instead of first calling a method to offset/shrink the alloc and then just having no positional arguments to read_scalar

Oh, interesting idea; I had not considered that.

Reply:
> Or does `offset` change the start without changing the end (i.e., it changes both start and size, the way things are currently represented)?

yea, this. Could probably use a better name